radix-engine/src/blueprints/resource/non_fungible/non_fungible_resource_manager.rs

```diff
@@ -96,6 +106,90 @@ where
 pub struct NonFungibleResourceManagerBlueprint;
 
 impl NonFungibleResourceManagerBlueprint {
+    fn validate_non_fungible_schema(
+        non_fungible_schema: &NonFungibleDataSchema,
+    ) -> Result<IndexMap<String, usize>, RuntimeError> {
+        let mut mutable_field_index = indexmap!();
+
+        // Validate schema
+        validate_schema(&non_fungible_schema.schema).map_err(|e| {
+            RuntimeError::ApplicationError(ApplicationError::NonFungibleResourceManagerError(
+                NonFungibleResourceManagerError::InvalidNonFungibleSchema(
+                    InvalidNonFungibleSchema::SchemaValidationError(e),
+                ),
+            ))
+        })?;
+
+        // Validate type kind
+        let type_kind = non_fungible_schema
+            .schema
+            .resolve_type_kind(non_fungible_schema.non_fungible)
+            .ok_or(RuntimeError::ApplicationError(
+                ApplicationError::NonFungibleResourceManagerError(
+                    NonFungibleResourceManagerError::InvalidNonFungibleSchema(
+                        InvalidNonFungibleSchema::InvalidLocalTypeIndex,
+                    ),
+                ),
+            ))?;
+
+        if !matches!(type_kind, TypeKind::Tuple { .. }) {
+            return Err(RuntimeError::ApplicationError(
+                ApplicationError::NonFungibleResourceManagerError(
+                    NonFungibleResourceManagerError::InvalidNonFungibleSchema(
+                        InvalidNonFungibleSchema::NotATuple,
+                    ),
+                ),
+            ));
+        }
+
+        // Validate names
+        let type_metadata = non_fungible_schema
+            .schema
+            .resolve_type_metadata(non_fungible_schema.non_fungible)
+            .ok_or(RuntimeError::ApplicationError(
+                ApplicationError::NonFungibleResourceManagerError(
+                    NonFungibleResourceManagerError::InvalidNonFungibleSchema(
+                        InvalidNonFungibleSchema::InvalidLocalTypeIndex,
+                    ),
+                ),
+            ))?;
+        match &type_metadata.child_names {
+            Some(ChildNames::NamedFields(names)) => {
+                let allowed_names: IndexMap<_, _> = names
+                    .iter()
+                    .enumerate()
+                    .map(|(i, x)| (x.as_ref(), i))
+                    .collect();
+                for f in &non_fungible_schema.mutable_fields {
+                    if let Some(index) = allowed_names.get(f.as_str()) {
+                        mutable_field_index.insert(f.to_string(), *index);
+                    } else {
+                        return Err(RuntimeError::ApplicationError(
+                            ApplicationError::NonFungibleResourceManagerError(
+                                NonFungibleResourceManagerError::InvalidNonFungibleSchema(
+                                    InvalidNonFungibleSchema::InvalidMutableField(f.to_string()),
```

Comment:
Can we change error to `MutableFieldDoesNotExist`

Reply:
Yes!